tests/test_reusable_executor.py

```diff
@@ -486,6 +488,46 @@ def test_call_item_gc_crash_or_exit(self, bad_object, match):
             for _ in executor.map(sleep, [.1] * 100):
                 pass
 
+    def test_sigkill_shutdown_leaks_workers(self):
+        # Create a parent process which will report its workers pids
+        # and sigkill itself.
+        code = """if True:
+        import os
+
+        import loky
+        import psutil
+
+        parent_pid = os.getpid()
+        with loky.get_reusable_executor(timeout=1, kill_workers=True) as p:
+            list(p.map(lambda x: x, range(100)))
+            for pid in p._processes.keys():
+                print(f'worker_pid:{pid}')
+            print(f'parent_pid:{parent_pid}')
+            psutil.Process(os.getpid()).kill()
```

Comment:
After looking back at the test, I don't think that this test will trigger the deadlock that is resolved in this PR.
Actually, as the children will only exit after ~30s in the case there is a deadlock, this test should fail.

I think to trigger the deadlock, you need long running tasks (`time.sleep` for instance) and then you need to have a wait mechanism for 30s in the test assert. WDYT?

Reply:
Actually the test just hangs with the fix removed:

```bash
diff --git a/loky/process_executor.py b/loky/process_executor.py
--- a/loky/process_executor.py	(revision 985d67141f376740308ffb1da5631320691986cb)
+++ b/loky/process_executor.py	(date 1650957151250)
@@ -411,12 +411,9 @@
         if call_item is None:
             # Notify queue management thread about worker shutdown
             result_queue.put(pid)
-            is_clean = worker_exit_lock.acquire(True, timeout=30)
-            if is_clean:
+            with worker_exit_lock:
                 mp.util.debug('Exited cleanly')
-            else:
-                mp.util.debug('Main process did not release worker_exit')
-            return
+                return
         try:
             r = call_item()
         except BaseException as e:
```
The test hangs indefinitely and loky processes still hang in the background:
```bash
$ pgrep -fla loky
99992 /home/haim0n/anaconda3/envs/loky_dev/bin/python -B -c from loky.backend.resource_tracker import main; main(12, False)
99998 /home/haim0n/anaconda3/envs/loky_dev/bin/python -B -c from loky.backend.resource_tracker import main; main(4, False)
99999 /home/haim0n/anaconda3/envs/loky_dev/bin/python -m loky.backend.popen_loky_posix --process-name LokyProcess-1 --pipe 13
100000 /home/haim0n/anaconda3/envs/loky_dev/bin/python -m loky.backend.popen_loky_posix --process-name LokyProcess-2 --pipe 14
100001 /home/haim0n/anaconda3/envs/loky_dev/bin/python -m loky.backend.popen_loky_posix --process-name LokyProcess-3 --pipe 15
100002 /home/haim0n/anaconda3/envs/loky_dev/bin/python -m loky.backend.popen_loky_posix --process-name LokyProcess-4 --pipe 16
100003 /home/haim0n/anaconda3/envs/loky_dev/bin/python -m loky.backend.popen_loky_posix --process-name LokyProcess-5 --pipe 17
100004 /home/haim0n/anaconda3/envs/loky_dev/bin/python -m loky.backend.popen_loky_posix --process-name LokyProcess-6 --pipe 18
100005 /home/haim0n/anaconda3/envs/loky_dev/bin/python -m loky.backend.popen_loky_posix --process-name LokyProcess-7 --pipe 19
100006 /home/haim0n/anaconda3/envs/loky_dev/bin/python -m loky.backend.popen_loky_posix --process-name LokyProcess-8 --pipe 20
```
I guess that's not a nice experience in case the bug still exits. 
Tried to mark the test as failure after given timeout with `@pytest.mark.timeout` , still no go.

Reply:
Ok sorry I should have run the test before. Indeed, this test is proper because the `subprocess.run` hangs while all the child workers are not stopped (that I missed).
Adding a timeout in the call make it so it does not hang when the test fails.